lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -239,16 +239,101 @@ OpFoldResult ShlOp::fold(ArrayRef<Attribute> operands) {
       operands, [](APInt a, APInt b) { return a.shl(b); });
 }
 
+LogicalResult ShlOp::canonicalize(ShlOp op, PatternRewriter &rewriter) {
+  // ShlOp(x, cst) -> Concat(Extract(x), zeros)
+  APInt value;
+  if (!matchPattern(op.rhs(), m_RConstant(value)))
+    return failure();
+
+  unsigned srcWidth = op.lhs().getType().cast<IntegerType>().getWidth();
+  unsigned shift = value.getZExtValue();
+
+  if (srcWidth <= shift) {
+    rewriter.replaceOpWithNewOp<hw::ConstantOp>(op,
+                                                APInt::getNullValue(srcWidth));
```

Comment:
I think this can go in the `fold` method for shift which will make it more general.